lib/marisa/exception.cc

```diff
@@ -0,0 +1,21 @@
+#include "marisa/exception.h"
+
+namespace marisa {
+
+/**
+ * This is defined here for the following reasons.
+ *
+ * 1. Virtual destructors are best not defined in a header to be included in many compilation units.
+ * Defining it here in a single compilation unit prevents multiple vtables from being generated and causing
+ * a compiler to be confused as to which is the true vtable, like with full link time optimization.
+ * By reducing the number of vtable instances, you also reduce the size of your library or application.
+ * You can see warnings around this with the -Wweak-vtables compiler option. Such issues may not be seen till runtime.
+ * Using the default keyword on the destructor is the same as defining it in the header.
+ *
+ * 2. Exceptions can not be thrown from destructors. When an exception is thrown from a destructor
```

Comment:
This seems to be an explanation of why it's `noexcept`, not why it's defined here.

Reply:
True. Point 1 is mostly around why it's in this file. Point 2 is more about why I changed the signature. I can eliminate point 2 if you don't think it's helpful.

Reply:
Some source code static analysis tools complain about the throws usage on destructors.

Reply:
Seems like something to be explained in the PR, not at the site of every destructor definition.

Reply:
This is the only destructor in this project with this issue.

Reply:
I also wanted the reasoning to be located with the code so that it doesn't get reverted by someone else's changes.